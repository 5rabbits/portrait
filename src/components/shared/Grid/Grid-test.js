import { React, toJSON, mount } from 'test/helper'
import Grid from 'shared/Grid'

describe('Grid', () => {
  it('should render with minimum props', () => {
    const component = mount(<Grid />)

    expect(toJSON(component)).toMatchSnapshot()
  })
})
