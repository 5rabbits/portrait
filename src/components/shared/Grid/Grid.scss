@import '~core/variables';

.grid-component {
  display: flex;

  // Alignment
  &.grid-component--start-align   { align-items: flex-start; }
  &.grid-component--center-align  { align-items: center; }
  &.grid-component--end-align     { align-items: flex-end; }
  &.grid-component--stretch-align { align-items: stretch; }

  // Justify
  &.grid-component--start-justify   { justify-content: flex-start; }
  &.grid-component--center-justify  { justify-content: center; }
  &.grid-component--end-justify     { justify-content: flex-end; }
  &.grid-component--around-justify  { justify-content: space-around; }
  &.grid-component--between-justify { justify-content: space-between; }

  // Direction
  &.grid-component--vertical-direction   { flex-direction: column; }
  &.grid-component--horizontal-direction { flex-direction: row; }

  // Gutter
  @mixin gutter($name, $spacing) {
    &.grid-component--#{$name}-gutter.grid-component--vertical-direction > * + * {
      margin-top: $spacing;
    }

    &.grid-component--#{$name}-gutter.grid-component--horizontal-direction > * + * {
      margin-left: $spacing;
    }
  }

  @include gutter('small', $spacing-base);
  @include gutter('normal', $spacing-base * 2);
  @include gutter('large', $spacing-base * 3);
}
