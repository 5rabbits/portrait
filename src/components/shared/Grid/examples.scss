@import '~core/variables';

.component-example--Grid {
  // Time entry example
  .time-entry {
    width: 200px;
    background-color: #4cb07d;
    color: #fff;
    border-radius: $border-radius-medium-component;
    padding: $spacing-base;
    font-size: $font-size-caption;
    line-height: $line-height-caption;

    > div:first-child {
      opacity: 0.6;
    }

    > div:last-child {
      font-size: 1rem;
    }
  }

  // Calendar example
  .calendar {
    $day-padding: $spacing-base * 2;

    height: 300px;
    background-color: #fff;
    width: 100%;

    .calendar-day__head {
      font-size: 0.8rem;
      padding: $day-padding;
    }

    .calendar-day__body {
      padding: 0 $day-padding $day-padding $day-padding;
    }

    .calendar-day__date {
      font-weight: bold;
    }

    .calendar-day__duration {
      color: $default-secondary-text-color;
    }

    .calendar-day + .calendar-day {
      border-left: $default-divider-color solid 1px;
    }

    .time-entry {
      width: auto;
    }
  }

  // Properties playground example
  .properties-playground {
    width: 100%;
    background-color: #fff;
    padding: $spacing-base * 2;

    .properties-playground__grid {
      margin-top: $spacing-base * 2;
      height: 400px;

      > div {
        background: #4cb07d;
        color: #fff;
        display: flex;
        align-items: center;
        justify-content: center;
        padding: $spacing-base;
      }
    }
  }
}
