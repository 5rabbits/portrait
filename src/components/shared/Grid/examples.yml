-
  description: |
    This example shows multiple Grid uses to build a calendar with time entries.

    **Note:** All `className` props are merely for styling, you can safely remove them to see only
    the layout placement.

  inline: false
  code: >
    const Calendar = () =>
      <Grid
        className="calendar"
        gutter={false}
        >
        <CalendarDay date="Monday 24" time="2h 25m">
          <TimeEntry height={65} />
          <TimeEntry height={100} />
        </CalendarDay>
        <CalendarDay date="Tuesday 25" time="0h 00m" />
        <CalendarDay date="Wednesday 26" time="3h 00m">
          <TimeEntry height={220} />
        </CalendarDay>
        <CalendarDay date="Thursday 27" time="0h 00m" />
        <CalendarDay date="Friday 28" time="0h 00m" />
      </Grid>

    const CalendarDay = ({ children, date, time, ...other }) =>
      <Grid
        {...other}
        className="calendar-day"
        direction="vertical"
        size={['min', 'fill']}
        gutter="small"
        >
        <Grid
          className="calendar-day__head"
          size={['fill', 'min']}
          >
          <div className="calendar-day__date">{date}</div>
          <div className="calendar-day__duration">{time}</div>
        </Grid>
        <Grid
          className="calendar-day__body"
          direction="vertical"
          gutter="small"
          size="min"
          >
          {children}
        </Grid>
      </Grid>

    const TimeEntry = ({ height }) =>
      <Grid
        className="time-entry"
        direction="vertical"
        gutter="small"
        size={['min', 'fill', 'min']}
        style={{ height }}
        >
        <div>Project</div>
        <div className="overflow-hidden">
          Aliqua culpa consequat mollit sint duis ullamco deserunt duis. Irure consequat velit est officia anim tempor.
        </div>
        <Grid
          align="center"
          size={['fill', 'min']}
          >
          <div>00h 00m</div>
          <i className="fa fa-lock" />
        </Grid>
      </Grid>

    ReactDOM.render(<Calendar />, mountNode)

-
  description: Properties playground.
  inline: false
  collapsible: true
  expanded: false
  code: >
    class Example extends React.Component {
      state = {
        align: 'center',
        direction: 'horizontal',
        gutter: 'normal',
        justify: 'center',
        size: 'min',
      }

      handleChange = (field, event) => {
        this.setState({
          [field]: event.target.value,
        })
      }

      render() {
        return (
          <div className="properties-playground">
            <form>
              <Grid size="min">
                <div>
                  <label>Align:</label>
                  {' '}
                  <select onChange={this.handleChange.bind(null, 'align')} value={this.state.align}>
                    <option value="start">Start</option>
                    <option value="center">Center</option>
                    <option value="end">End</option>
                    <option value="stretch">Stretch</option>
                  </select>
                </div>
                <div>
                  <label>Direction:</label>
                  {' '}
                  <select onChange={this.handleChange.bind(null, 'direction')} value={this.state.direction}>
                    <option value="horizontal">Horizontal</option>
                    <option value="vertical">Vertical</option>
                  </select>
                </div>
                <div>
                  <label>Gutter:</label>
                  {' '}
                  <select onChange={this.handleChange.bind(null, 'gutter')} value={this.state.gutter}>
                    <option value={false}>None</option>
                    <option value="small">Small</option>
                    <option value="normal">Normal</option>
                    <option value="large">Large</option>
                  </select>
                </div>
                <div>
                  <label>Size:</label>
                  {' '}
                  <select onChange={this.handleChange.bind(null, 'size')} value={this.state.size}>
                    <option value="fill">Fill</option>
                    <option value="min">Min</option>
                  </select>
                </div>
                <div>
                  <label>Justify:</label>
                  {' '}
                  <select onChange={this.handleChange.bind(null, 'justify')} value={this.state.justify}>
                    <option value="start">Start</option>
                    <option value="center">Center</option>
                    <option value="end">End</option>
                    <option value="around">Around</option>
                    <option value="between">Between</option>
                  </select>
                </div>
              </Grid>
            </form>

            <Grid {...this.state} className="properties-playground__grid">
              <div style={{ minHeight: 60 }}>1</div>
              <div style={{ minHeight: 40 }}>2</div>
              <div style={{ minHeight: 50 }}>3</div>
              <div style={{ minHeight: 70 }}>4</div>
              <div style={{ minHeight: 40 }}>5</div>
            </Grid>
          </div>
        )
      }
    }

    ReactDOM.render(<Example />, mountNode)
