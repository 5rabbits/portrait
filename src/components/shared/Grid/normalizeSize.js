export default size => {
  switch (size) {
    case null:
      return undefined

    case undefined:
      return undefined

    case 'fill':
      return 1

    case 'min':
      return '0 0 auto'

    default:
      return size
  }
}
