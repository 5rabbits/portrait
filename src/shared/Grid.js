import React from 'react'

const Grid = () =>
  <div className="container-page">
    {this.props.children}
  </div>

export default Grid
