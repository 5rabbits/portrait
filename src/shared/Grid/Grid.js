import React from 'react'
import childrenPropType from 'propTypes/children'

// NO USAR

const Grid = props =>
  <div className="container-page">
    {props.children}
  </div>

Grid.propTypes = {
  children: childrenPropType,
}

export default Grid
