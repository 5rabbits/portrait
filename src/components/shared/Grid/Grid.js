import React, { PropTypes, Children } from 'react'
import childrenPropType from 'propTypes/children'
import cx from 'classnames'
import './Grid.scss'
import normalizeSize from './normalizeSize'

/**
 * Allows to place elements using the [CSS Flexible Box Layout](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Flexible_Box_Layout).
 */
const Grid = ({ align, children, className, direction, gutter, justify, size, ...other }) => {
  const childrenArray = Children.toArray(children)
  const columns = []

  for (let i = 0; i < childrenArray.length; i += 1) {
    columns.push(React.cloneElement(childrenArray[i], {
      style: {
        ...(childrenArray[i].props.style || {}),
        flex: normalizeSize(size instanceof Array ? size[i] : size),
      },
    }))
  }

  return (
    <div
      {...other}
      className={cx('grid-component', className, {
        [`grid-component--${align}-align`]: align,
        [`grid-component--${direction}-direction`]: direction,
        [`grid-component--${gutter}-gutter`]: gutter,
        [`grid-component--${justify}-justify`]: justify,
      })}
      >
      {columns}
    </div>
  )
}

Grid.propTypes = {
  /**
   * Changes the alignment on the opposite axis.
   *
   * For example, `center` will center items vertically if `direction="horizontal"` and
   * horizontally if `direction="vertical"`.
   */
  align: PropTypes.oneOf(['start', 'center', 'end', 'stretch']).isRequired,

  /**
   * Any number of items that will be placed in the grid.
   */
  children: childrenPropType,

  /**
   * CSS class of the grid outer wrapper.
   */
  className: PropTypes.string,

  /**
   * Direction in which the items are going to be placed.
   *
   * The items are going to be placed left to right if `horizontal` and top to bottom
   * if `vertical`.
   */
  direction: PropTypes.oneOf(['vertical', 'horizontal']).isRequired,

  /**
   * Space between items.
   */
  gutter: PropTypes.oneOf([false, 'small', 'normal', 'large']),

  /**
   * Manages negative space on the main axis.
   *
   * For example, `center` will center items vertically if `direction="vertical"` and
   * horizontally if `direction="horizontal"`.
   */
  justify: PropTypes.oneOf(['start', 'center', 'end', 'between', 'around']).isRequired,

  /**
   * Size of the items.
   *
   * A size unit correspond any of the following values.
   * - `'fill'`: Distribute the available space equally between all items with `fill`.
   * - `'min'`: Use the minimum space possible.
   * - Using any *string* or *number* will pass the value directly to the css
   * [flex](https://developer.mozilla.org/en-US/docs/Web/CSS/flex) property.
   *
   * The size can be specified globally (`size="fill"`) or item by item (`size={['fill', 'min']}`).
   */
  size: PropTypes.oneOfType([
    PropTypes.oneOf(['fill', 'min']),
    PropTypes.string,
    PropTypes.number,
    PropTypes.arrayOf(PropTypes.oneOfType([
      PropTypes.number, PropTypes.string,
    ])),
  ]),
}

Grid.defaultProps = {
  align: 'stretch',
  direction: 'horizontal',
  gutter: 'normal',
  justify: 'start',
  size: 'fill',
}

export default Grid
