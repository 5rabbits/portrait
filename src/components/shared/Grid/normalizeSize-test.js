import normalizeSize from 'shared/Grid/normalizeSize'

describe('normalizeSize(size)', () => {
  describe('if size is null', () => {
    it('should return undefined', () => {
      expect(normalizeSize(null)).toBeUndefined()
    })
  })

  describe('if size is undefined', () => {
    it('should return undefined', () => {
      expect(normalizeSize()).toBeUndefined()
    })
  })

  describe('if size is "fill"', () => {
    it('should return 1', () => {
      expect(normalizeSize('fill')).toBe(1)
    })
  })

  describe('if size is "min"', () => {
    it('should return "0 0 auto"', () => {
      expect(normalizeSize('min')).toBe('0 0 auto')
    })
  })

  describe('if size is other value', () => {
    it('should return size directly', () => {
      expect(normalizeSize('1 1 100%')).toBe('1 1 100%')
    })
  })
})
