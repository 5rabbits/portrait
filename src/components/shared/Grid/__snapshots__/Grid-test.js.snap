// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid should render with minimum props 1`] = `
<Grid
  align="stretch"
  direction="horizontal"
  gutter="normal"
  justify="start"
  size="fill"
>
  <div
    className="grid-component grid-component--stretch-align grid-component--horizontal-direction grid-component--normal-gutter grid-component--start-justify"
  />
</Grid>
`;
